File: jupyter/nb/source/pl08_minc/jl/minc/minc_ast.jl
import Printf

#= 
Abstract Syntax Tree 
=#

#= type expression:

   for now, we only have a primitive type (long),
   which is TypePrimitive("long")
=#

abstract type TypeExpr end

struct TypePrimitive <: TypeExpr
    name :: String              # type name (always "long", for now)
end

#= variable declaration:

   long x; -> Decl(TypePrimitive("long"), "x")
   also (ab)used to represent a function parameter
=#
struct Decl
    var_type :: TypeExpr        # variable type
    name :: String              # variable name
end

# expression
abstract type Expr end

# 1, 2, 3, ... 
struct ExprIntLiteral <: Expr
    val :: Int64
end

# x, y, z, ...
struct ExprId <: Expr
    name :: String
end

# -x, x - y, ...
struct ExprOp <: Expr
    op :: String
    args :: Vector{Expr}
end

# f(1, 2, 3)
struct ExprCall <: Expr
    fun :: Expr
    args :: Vector{Expr}
end

# (x + y)
struct ExprParen <: Expr
    sub_expr :: Expr
end

#= 
statement
=#
abstract type Stmt end

# ;
struct StmtEmpty <: Stmt end

# continue;
struct StmtContinue <: Stmt end

# break;
struct StmtBreak <: Stmt end

# return e;
struct StmtReturn <: Stmt
    expr :: Expr
end

# f(x);
struct StmtExpr <: Stmt
    expr :: Expr
end

# { int x; return x + 1; } 
struct StmtCompound <: Stmt
    decls :: Vector{Decl}
    stmts :: Vector{Stmt}
end    

# if (expr) stmt [else stmt]
struct StmtIf <: Stmt
    cond :: Expr
    then_stmt :: Stmt
    else_stmt :: Union{Stmt,Nothing}
end

# while (expr) stmt
struct StmtWhile <: Stmt
    cond :: Expr
    body :: Stmt
end

#= 
toplevel definition
=#
abstract type Def end

#= 
function definition
e.g., long f(long x, long y) { return x; }
=#
struct DefFun <: Def
    name :: String
    params :: Vector{Decl}
    return_type :: TypeExpr
    body :: Stmt
end

# program is just a list of definitions
struct Program
    defs :: Vector{Def}
end

#=
convert ast back to C string

ast_to_str_xxx converts an AST to a string 
valid as a C program.

it will not be used anywhere in your compiler, 
but is given for illustrating how you walk
the AST and what kind of C program each AST 
is actually meant to represent.
=#

#=
 AST for type expression -> C string
TypeExpr::Primitive{"long"} -> "long"
=#
function ast_to_str_type(type_expr :: TypePrimitive)
    type_expr.name
end

#= 
AST for function parameter -> C string
Decl{TypePrimitive("long"), "x"} -> "long x"
=#
function ast_to_str_param(param :: Decl)
    Printf.@sprintf("%s %s", ast_to_str_type(param.var_type), param.name)
end

#=
 AST for a variable declaration -> C string
Decl{TypePrimitive("long"), "x"} -> "long x;"
=#
function ast_to_str_decl(decl :: Decl)
    Printf.@sprintf("%s %s;", decl.var_type, decl.name)
end

#= 
AST for an expression -> C string
=#

# ExprIntLiteral{123} -> "123"
function ast_to_str_expr(expr :: ExprIntLiteral)
    Printf.@sprintf("%d", expr.val)
end

# ExprId("x") -> "x"
function ast_to_str_expr(expr :: ExprId)
    expr.name
end

# ExprOp("+" [ExprId("x"); ExprIntLiteral("123")]) -> "x + 123"
function ast_to_str_expr(expr :: ExprOp)
    return join(map(ast_to_str_expr, expr.args), 
                Printf.@sprintf(" %s ", expr.op))
end

# ExprCall(ExprId("f"), [ExprId("x"); ExprIntLiteral("123")]) -> "f(x, 123)"
function ast_to_str_expr(expr :: ExprCall)
    fun = ast_to_str_expr(expr.fun)
    args = join(map(ast_to_str_expr, expr.args), ", ")
    Printf.@sprintf("%s(%s)", fun, args)
end

# ExprParen(ExprOp("+", [ExprId("x"); ExprIntLiteral("123")])) -> "(x + 123)"
function ast_to_str_expr(expr :: ExprParen)
    sub_expr = ast_to_str_expr(expr.sub_expr)
    Printf.@sprintf("(%s)", sub_expr)
end

#=
AST for a statement -> C string
=#

function ast_to_str_stmt(stmt :: StmtEmpty)
    ";"
end

function ast_to_str_stmt(stmt :: StmtContinue)
    "continue;"
end

function ast_to_str_stmt(stmt :: StmtBreak)
    "break;"
end

function ast_to_str_stmt(stmt :: StmtReturn)
    Printf.@sprintf("return %s;", ast_to_str_expr(stmt.expr))
end

function ast_to_str_stmt(stmt :: StmtExpr)
    Printf.@sprintf("%s;", ast_to_str_expr(stmt.expr))
end

function ast_to_str_stmt(stmt :: StmtCompound)
    decls = join(map(ast_to_str_decl, stmt.decls), "\n")
    stmts = join(map(ast_to_str_stmt, stmt.stmts), "\n")
    Printf.@sprintf("{\n%s\n%s\n}", decls, stmts)
end

function ast_to_str_stmt(stmt :: StmtIf)
    cond = ast_to_str_expr(stmt.cond)
    then_stmt = ast_to_str_stmt(stmt.then_stmt)
    if stmt.else_stmt == nothing
	Printf.@sprintf("if (%s) %s", cond, then_stmt)
    else
	else_stmt = ast_to_str_stmt(stmt.else_stmt)
	Printf.@sprintf("if (%s) %s else %s", cond, then_stmt, else_stmt)
    end
end

function ast_to_str_stmt(stmt :: StmtWhile)
    cond = ast_to_str_expr(stmt.cond)
    body = ast_to_str_stmt(stmt.body)
    Printf.@sprintf("while (%s) %s", cond, body)
end

#=
 AST for a definition -> C string


   DefFun("f", [Decl(TypePrimitive("long"), "x")], TypePrimitive("long"), 
                    StmtCompound([], [])) ->
       "long f(long x) {}"
=#
function ast_to_str_def(def :: DefFun)
    fname = def.name
    params = join(map(ast_to_str_param, def.params), ", ")
    return_type = ast_to_str_type(def.return_type)
    body = ast_to_str_stmt(def.body)
    Printf.@sprintf("%s %s(%s) %s", return_type, fname, params, body)
end

function ast_to_str_program(prog :: Program)
    join(map(ast_to_str_def, prog.defs), "\n")
end

File: jupyter/nb/source/pl08_minc/rs/minc/src/minc_ast.rs
/* Abstract Syntax Tree */

/* type expression:

   for now, we only have a primitive type (long),
   which is TypeExpr::Primitive("long") */

pub enum TypeExpr {
    Primitive(String)		// type name (always "long", for now)
}

/* variable declaration:

   long x; -> Decl{var_type=TypeExpr::Primitive("long"), name="x"}
   also (ab)used to represent a function parameter */

pub struct Decl {
    pub var_type : TypeExpr,
    pub name : String
}

/* expression */
pub enum Expr {
    IntLiteral(i64),            /* 1, 2, 3, ... */
    Id(String),                 /* x, y, z, ... */
    Op(String, Vec<Expr>),      /* -x, x - y, ... */
    Call(Box<Expr>, Vec<Expr>), /* f(1, 2, 3) */
    Paren(Box<Expr>)            /* (x + y) */
}

/* statement */

pub enum Stmt {
    Empty,                      /* ; */
    Continue,                   /* continue; */
    Break,                      /* break; */
    Return(Expr),               /* return e; */
    Expr(Expr),                 /* f(x); */
    Compound(Vec<Decl>, Vec<Stmt>), /* { int x; return x + 1; } */
    If(Expr, Box<Stmt>, Option<Box<Stmt>>), /* if (expr) stmt [else stmt] */
    While(Expr, Box<Stmt>)                  /* while (expr) stmt */
}

/* toplevel definition */
pub enum Def {
    /* function definition
     e.g., long f(long x, long y) { return x; } */
    Fun(String, Vec<Decl>, TypeExpr, Stmt)
}

/* program is just a list of definitions */
pub struct Program {
    pub defs : Vec<Def>
}

/* convert ast back to C string 

   ast_to_str_xxx converts an AST to a string 
   valid as a C program.

   it will not be used anywhere in your compiler, 
   but is given for illustrating how you walk
   the AST and what kind of C program each AST 
   is actually meant to represent.
*/

/* AST for type expression -> C string
   TypePrimitive{"long"} -> "long" */
#[allow(dead_code)]
pub fn ast_to_str_type(type_expr : &TypeExpr) -> String {
    match type_expr {
        TypeExpr::Primitive(name) => { format!("{}", name) }
    }
}

/* AST for function parameter -> C string
   Decl{TypeExpr::Primitive("long"), "x"} -> "long x" */
#[allow(dead_code)]
pub fn ast_to_str_param(param : &Decl) -> String {
    format!("{} {}", ast_to_str_type(&param.var_type), param.name)
}

/* AST for a variable declaration -> C string
   Decl{TypeExpr::Primitive("long"), "x"} -> "long x;" */
#[allow(dead_code)]
pub fn ast_to_str_decl(decl : &Decl) -> String {
    format!("{} {};", ast_to_str_type(&decl.var_type), decl.name)
}

/* AST for an expression -> C string */
#[allow(dead_code)]
pub fn ast_to_str_expr(expr : &Expr) -> String {
    match expr {
        /* Expr::IntLiteral{123} -> "123" */
        Expr::IntLiteral(val) => format!("{}", val),
        /* Expr::Id("x") -> "x" */
        Expr::Id(name) => format!("{}", name),
        /* Expr::Op("+" [Expr::Id("x"); Expr::IntLiteral("123")]) -> "x + 123" */
        Expr::Op(op, args) => {
            args.iter().map(ast_to_str_expr).collect::<Vec<String>>().join(&format!(" {} ", op))
        },
        /* Expr::Call(Expr::Id("f"), [Expr::Id("x"); Expr::IntLiteral("123")]) -> "f(x, 123)" */
        Expr::Call(fun, args) => {
            format!("{}({})",
                    ast_to_str_expr(fun),
                    args.iter().map(ast_to_str_expr).collect::<Vec<String>>().join(", "))
        },
        /* Expr::Paren(Expr::Op("+", [Expr::Id("x"); Expr::IntLiteral("123")])) -> "(x + 123)" */
        Expr::Paren(sub_expr) => {
            format!("({})", ast_to_str_expr(sub_expr))
        }
    }
}

/* AST for a statement -> C string */
#[allow(dead_code)]
pub fn ast_to_str_stmt(stmt : &Stmt) -> String {
    match stmt {
        Stmt::Empty => format!(";"),
        Stmt::Continue => format!("continue;"),
        Stmt::Break => format!("break;"),
        Stmt::Return(expr) => format!("return {};", ast_to_str_expr(expr)),
        Stmt::Expr(expr) => format!("{};", ast_to_str_expr(expr)),
        Stmt::Compound(decls, stmts) => {
            format!("{{\n{}\n{}\n}}", 
                    decls.iter().map(ast_to_str_decl).collect::<Vec<String>>().join("\n"),
                    stmts.iter().map(ast_to_str_stmt).collect::<Vec<String>>().join("\n"))
        },
        Stmt::If(cond, then_stmt, Some(else_stmt)) => {
            format!("if ({}) {} else {}",
                    ast_to_str_expr(cond),
                    ast_to_str_stmt(then_stmt),
                    ast_to_str_stmt(else_stmt))
        }
        Stmt::If(cond, then_stmt, None) => {
            format!("if ({}) {}",
                    ast_to_str_expr(cond),
                    ast_to_str_stmt(then_stmt))
        }
        Stmt::While(cond, body) => {
            format!("while ({}) {}",
                    ast_to_str_expr(cond),
                    ast_to_str_stmt(body))
        }
    }
}

/* AST for a definition -> C string

   DefFun("f", [Decl(TypeExpr::Primitive("long"), "x")], TypeExpr::Primitive("long"), 
                    StmtCompound([], [])) ->
       "long f(long x) {}"
*/
#[allow(dead_code)]
pub fn ast_to_str_def(def : &Def) -> String {
    match def {
        Def::Fun(name, params, return_type, body) => {
            format!("{} {}({}) {}",
                    ast_to_str_type(return_type),
                    name,
                    params.iter().map(ast_to_str_param).collect::<Vec<String>>().join(", "),
                    ast_to_str_stmt(body))
        }
    }
}

#[allow(dead_code)]
pub fn ast_to_str_program(program : &Program) -> String {
    program.defs.iter().map(ast_to_str_def).collect::<Vec<String>>().join("\n")
}

